### ISSUE TITLE
Incorrect complain about async function return value type

### ISSUE THREAD

**Issue by serhiy-storchaka** [python/mypy#7736]
The following example
```python
def t() -> None:
    async def f() -> int:
        return 1

    def g() -> int:
        return len(x)

    x = [1]
```
produces an incorrect MyPy error:
```
coro.py:3: error: Incompatible return value type (got "int", expected "Coroutine[Any, Any, int]")
```
At module level all works. If remove `g` all works. If make `g` returning a constant all works. If move the definitions of `g` after `x` all works.

**Comment by asvetlov** [comment]
Thanks for filing an issue.
I've started to encounter this problem since mypy 0.720 but was unable to reproduce in a simple test.
I did not figure out that nested function-level scope is needed.

**Comment by ilevkivskyi** [comment]
Thanks for reporting! This looks like a bug in the new semantic analyzer. The function gets deferred (because there is a forward reference) so its return type is wrapped into `Coroutine[...]` twice.

There is some logic that tries to guard against this (wrapping happens conditionally under `if ... and not self.deferred: ...`), but the flag gets reset between the two iterations.
